sdk-extensions/jaeger-remote-sampler/src/main/java/io/opentelemetry/sdk/extension/trace/jaeger/sampler/DefaultGrpcServiceBuilder.java

```diff
@@ -0,0 +1,170 @@
+
+import io.grpc.Codec;
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.Metadata;
+import io.grpc.stub.MetadataUtils;
+import io.opentelemetry.api.metrics.MeterProvider;
+import io.opentelemetry.exporter.otlp.internal.Marshaler;
+import io.opentelemetry.exporter.otlp.internal.grpc.ManagedChannelUtil;
+import io.opentelemetry.exporter.otlp.internal.grpc.MarshalerServiceStub;
+import io.opentelemetry.exporter.otlp.internal.grpc.OkHttpGrpcExporterBuilder;
+import io.opentelemetry.exporter.otlp.internal.retry.RetryPolicy;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLException;
+
+final class DefaultGrpcServiceBuilder<ReqT extends Marshaler, ResT extends UnMarshaller>
+    implements GrpcServiceBuilder<ReqT, ResT> {
+
+  private final String type;
+  private final Function<ManagedChannel, MarshalerServiceStub<ReqT, ResT, ?>> stubFactory;
+  private final String grpcServiceName;
+
+  @Nullable private ManagedChannel channel;
+  private long timeoutNanos;
+  private URI endpoint;
+  private boolean compressionEnabled = false;
+  @Nullable private Metadata metadata;
+  @Nullable private byte[] trustedCertificatesPem;
+  @Nullable private RetryPolicy retryPolicy;
+
+  /** Creates a new {@link OkHttpGrpcExporterBuilder}. */
+  // Visible for testing
+  DefaultGrpcServiceBuilder(
+      String type,
+      Function<ManagedChannel, MarshalerServiceStub<ReqT, ResT, ?>> stubFactory,
+      long defaultTimeoutSecs,
+      URI defaultEndpoint,
+      String grpcServiceName) {
+    this.type = type;
+    this.stubFactory = stubFactory;
+    this.grpcServiceName = grpcServiceName;
+    timeoutNanos = TimeUnit.SECONDS.toNanos(defaultTimeoutSecs);
+    endpoint = defaultEndpoint;
+  }
+
+  @Override
+  public DefaultGrpcServiceBuilder<ReqT, ResT> setChannel(ManagedChannel channel) {
+    this.channel = channel;
+    return this;
+  }
+
+  @Override
+  public DefaultGrpcServiceBuilder<ReqT, ResT> setTimeout(long timeout, TimeUnit unit) {
+    timeoutNanos = unit.toNanos(timeout);
+    return this;
+  }
+
+  @Override
+  public DefaultGrpcServiceBuilder<ReqT, ResT> setTimeout(Duration timeout) {
+    return setTimeout(timeout.toNanos(), TimeUnit.NANOSECONDS);
+  }
+
+  @Override
+  public DefaultGrpcServiceBuilder<ReqT, ResT> setEndpoint(String endpoint) {
+    URI uri;
+    try {
+      uri = new URI(endpoint);
+    } catch (URISyntaxException e) {
+      throw new IllegalArgumentException("Invalid endpoint, must be a URL: " + endpoint, e);
+    }
+
+    if (uri.getScheme() == null
+        || (!uri.getScheme().equals("http") && !uri.getScheme().equals("https"))) {
+      throw new IllegalArgumentException(
+          "Invalid endpoint, must start with http:// or https://: " + uri);
+    }
```

Comment:
Let's extract this to a helper function now that it appears in a couple of places.

Reply:
done